controllers/clustertopology_controller_compute.go

```diff
@@ -78,8 +84,286 @@ func (r *ClusterTopologyReconciler) getCurrentState(ctx context.Context, cluster
+	if cluster.Labels == nil {
+		cluster.Labels = map[string]string{}
+	}
+	cluster.Labels[clusterv1.ClusterLabelName] = cluster.Name
+	cluster.Labels[clusterv1.ClusterTopologyLabelName] = ""
+
+	// Set the references to the infrastructureCluster and controlPlane objects.
+	// NOTE: Once set for the first time, the references are not expected to change.
+	cluster.Spec.InfrastructureRef = objToRef(infrastructureCluster)
+	cluster.Spec.ControlPlaneRef = objToRef(controlPlane)
+
+	return cluster
+}
+
+type templateToInput struct {
+	template              *unstructured.Unstructured
+	templateClonedFromRef *corev1.ObjectReference
+	cluster               *clusterv1.Cluster
+	namePrefix            string
+	currentObjectRef      *corev1.ObjectReference
+	labels                map[string]string
+	annotations           map[string]string
+}
+
+// templateToObject generates an object from a template, taking care
+// of adding required labels (cluster, topology), annotations (clonedFrom)
+// and assigning a meaningful name (or reusing current reference name).
+func templateToObject(in templateToInput) (*unstructured.Unstructured, error) {
+	// Enforce the topology labels into the provided label set.
+	// NOTE: The cluster label is added at creation time so this object could be read by the ClusterTopology
+	// controller immediately after creation, even before other controllers are going to add the label (if missing).
+	labels := in.labels
+	if labels == nil {
+		labels = map[string]string{}
+	}
+	labels[clusterv1.ClusterLabelName] = in.cluster.Name
+	labels[clusterv1.ClusterTopologyLabelName] = ""
+
+	// Generate the object from the template.
+	// NOTE: OwnerRef can't be set at this stage; other controllers are going to add OwnerReferences when
+	// the object is actually created.
+	object, err := external.GenerateTemplate(&external.GenerateTemplateInput{
+		Template:    in.template,
+		TemplateRef: in.templateClonedFromRef,
+		Namespace:   in.cluster.Namespace,
+		Labels:      labels,
+		Annotations: in.annotations,
+		ClusterName: in.cluster.Name,
+	})
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to compute desired state for the cluster infrastructure object")
+	}
+
+	// Ensure the generated objects have a meaningful name.
+	// NOTE: In case there is already a ref to this object in the Cluster, re-use the same name
+	// in order to simplify compare at later stages of the reconcile process.
+	object.SetName(names.SimpleNameGenerator.GenerateName(in.namePrefix))
+	if in.currentObjectRef != nil && len(in.currentObjectRef.Name) > 0 {
+		object.SetName(in.currentObjectRef.Name)
+	}
+
+	return object, nil
+}
+
+// templateToTemplate generates a template from an existing template, taking care
+// of adding required labels (cluster, topology), annotations (clonedFrom)
+// and assigning a meaningful name (or reusing current reference name).
+// NOTE: We are creating a copy of the ClusterClass template for each cluster so
```

Comment:
```suggestion
// NOTE: We are creating a copy of the InfrastructureCluster template for each cluster so
```

Reply:
Although I'm not sure if this comment isn't to specific given that this func is used in multiple contexts